File: kattis/teque.py
from collections import deque


class teque:
    def __init__(self):
        self._head = deque()
        self._tail = deque()

    def __len__(self):
        return len(self._head) + len(self._tail)

    def __getitem__(self, i):
        if i < len(self._head):
            return self._head[i]
        else:
            return self._tail[i-len(self._head)]

    def _balance(self):
        while len(self._head) - 1 > len(self._tail):
            self._tail.appendleft(self._head.pop())
        while len(self._head) < len(self._tail):
            self._head.append(self._tail.popleft())

    def push_back(self, x):
        self._tail.append(x)

    def push_front(self, x):
        self._head.appendleft(x)

    def push_middle(self, x):
        self._balance()
        self._head.append(x)


tq = teque()
for _ in range(int(input())):
    c, x = input().split()
    x = int(x)
    if c == "push_back":
        tq.push_back(x)
    elif c == "push_front":
        tq.push_front(x)
    elif c == "push_middle":
        tq.push_middle(x)
    elif c == "get":
        print(tq[x])

File: kattis/teque.cc
#include <deque>

class teque
{

    std::deque<int> head;
    std::deque<int> tail;

public:

    void balance()
    {
        while (head.size() - 1 > tail.size())
        {
            tail.push_front(head.back());
            head.pop_back();
        }
        while (head.size() < tail.size())
        {
            head.push_back(tail.front());
            tail.pop_front();
        }
    }

    size_t size()
    {
        return head.size() + tail.size();
    }

    int &operator[](size_t i)
    {
        if (i < head.size())
            return head[i];
        else
            return tail[i - head.size()];
    }

    void push_back(int x)
    {
        tail.push_back(x);
    }

    void push_front(int x)
    {
        head.push_front(x);
    }

    void push_middle(int x)
    {
        balance();
        head.push_back(x);
    }
};

#include <cstdio>
#include <cstring>

int main()
{
    teque tq;
    unsigned n;
    char s[16];
    int x;
    std::scanf("%u", &n);
    for (unsigned i = 0; i < n; i++)
    {
        std::scanf("%s %u", s, &x);
        if (std::strcmp(s, "push_back") == 0)
            tq.push_back(x);
        else if (std::strcmp(s, "push_front") == 0)
            tq.push_front(x);
        else if (std::strcmp(s, "push_middle") == 0)
            tq.push_middle(x);
        else if (std::strcmp(s, "get") == 0)
            std::printf("%u\n", tq[x]);
    }
    return 0;
}
